[체스 - 3, 4단계] 조이썬(이영수) 미션 제출합니다.

youngsu5582 opened the pull request:
제이! 오랜만에 뵈는거 같습니다,,
DB를 연동한다는게 생각보다 되게 어렵네요.. 😭😭

그전과 비슷한 맥락으로
제가 구현을 어떤 방식으로 접근 했는지와 느낀점에 대해서 +
구현중 궁금했던 점들에 대해서 질문하는 방식으로 작성 하겠습니다!

---
### 느낀점
#### 1. Fake Repository 구현

4단계 미션을 진행하며
우선 interface 와 Map 을 활용하여 Fake Repository 를 먼저 구현한 후
실제 DB와 연동하는 Repository 를 만드는 식으로 구현을 하였습니다.

이유로는
1. 내가 필요한 데이터 , 메소드 가 뭔지 설계 & 동작하는 원리를 알기 위해서
2. 먼저 작성하고 통과한 테스트 코드를 기반으로 DB를 하나씩 확인 가능하기 위해서
가 있었습니다

구현하며 느낀점은
똑같은 작업이 반복은 되나 , 편안함을 느꼈습니다.
해당 내용을 기반으로 비즈니스 로직을 완료함 후,
디비로 교체하여 사용하니 문제없이 바로 교체가 가능했기 때문인데요,
실무에서도 이렇게 FakeRepository 를 통한 접근 방식이 자주 사용되는지 궁금합니다!

#### 2. Enum 과 DB 매칭

해당 미션을 구현하며 디비 설계를 했습니다.
그리고 , DB 에는 객체가 아닌 문자열을 저장할 수 밖에 없습니다.
DTO 를 변화시켜주듯이 Mapper 를 사용할까 생각했으나

다른 클래스에서 관리하는게 아닌 해당 값은 해당 값과 매칭 되는 Enum 이 가지고 있는게 더 적합하다 생각하여
```java
public enum Color {
    BLACK("black"),
    WHITE("white");

    public static final Color from(final String value) {
        return Arrays.stream(values())
                     .filter(color -> color.getValue()
                                           .equals(value))
                     .findFirst()
                     .orElseThrow(() -> new IllegalArgumentException(String.format("%s 은 Color 에 없습니다", value)));
    }
}
```
문자열 과 문자열을 통해 Enum을 반환하는 로직을 만들었습니다.
DB 의 값과 Enum 내 문자열을 매칭 시키는게 일반적인지 궁금합니다!

#### 3. GameEndException

3단계 점수 및 게임 종료 로직을 구현하며
왕이 잡히면 다음 입력 로직 이나 , 비즈니스 로직을 수행하지 않고
바로 종료가 되게 하고 싶었습니다.

그렇기에 `GameEndException` 같이 Exception 을 발생시키고 , catch 하는 방법을 사용했습니다.
이때 든 생각으로
객체지향의 세계에서는 결국 객체간 메시지들을 보내 단방향으로 오가야 하나
Exception 을 비즈니스 로직에 사용하여 의도적으로 로직을 끊는 일종의 C언어의 goto의 느낌을 받았습니다.

또 한편으론 , 체스 게임에선
왕이 잡히면 게임이 끝나는 것이니 GameEndException 이나 CheckMateException 을 발생 시켜도 괜찮지 않을까?
라는 생각도 들기도 합니다.

비즈니스 로직 & 로직 중 의도적으로 Exception 을 발생시켜 로직을 수행해도 괜찮은지 궁금합니다!
( 아니면 , Exception 을 통해 비즈니스 로직 & 게임을 종료 시켜도 되는지...! )

#### 4. getScore 시 , 불필요한 List<Piece> 전달

현재의 요구사항은
일반적으로 다른 기물 들은 단순 상수 점수값을 반환하나
폰은 같은 랭크 에 다른 아군 폰이 있을 시 0.5점을 반환하여 추가된 로직이 포함되어 있습니다
처음에는 폰은 특별하게 분리를 하여 로직을 구분을 할까 했지만

1. 다른 기물들 에서 이와 유사한 로직을 해야할 시 또 isXX 를 만들어 로직을 분리해야 한다는 점
2. Pawn 이 List<Piece> 구문이 필요 없어진다면 삭제하기 용이하다는 점

이러한 점들 때문에 사용하지 않지만 기물들 getScore 에 List<Piece> 를 넣었습니다!
위 접근 방식이 타당 할지 궁금합니다 제이!

---

이상으로 , 체스 구현을 요구사항에는 부합하게 맞췄는데
레벨1 이 끝나간다고 하니 아쉽기도 하며 , 후련하기도 한 거 같습니다!

마지막 까지 잘 부탁드립니다! 감사합니다!! ☺️ 🫡🫡

youngsu5582 committed:
feat(PieceCheckable) : 포인트에 피스가 있는지 확인 하는 기능 기술하는 interface 명세

youngsu5582 committed:
feat(PieceChecker) : PieceCheckable 을 구현하는 도메인 기능 구현

Pieces 와 동일한 로직이므로 테스트 코드 작성 X
- 이에 따른 각 Piece 들 변수 , 매개변수 변경

youngsu5582 committed:
feat(PieceChecekr) : 기존 Piece 에서 담당하는 로직들 이동

포인트가 같은 색깔인지
포인트가 같은 색깔이 아닌지
포인트가 비어 있는지
포인트 에서 포인트 까지 기물이 있는지

Pieces 와 동일한 로직이므로 테스트 코드 작성 X
- 이에 따른 각 Piece 들 변수 , 매개변수 변경

youngsu5582 committed:
refactor(Piece) : 기존 기물에서 이동함에 따라 각 기물들 변경

youngsu5582 committed:
feat(Pawn) : 파일 기반 점수 결정하는 기능 구현

youngsu5582 committed:
feat(Piece) : 각 기물당 점수 반환하는 기능 구현

youngsu5582 committed:
feat(Pieces) : 색깔에 다른 기물들 총 점수 반환 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
fix(Piece) : 메소드 잘못 들어간 부분 수정

youngsu5582 committed:
feat(ChessBoard) : 체스보드 생성시 킹 두개가 있는지 검증하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(All) : Controller , View 적용함으로 3단계 기능 구현 완료

- 완료함에 따라 기능 요구 사항 체크

youngsu5582 committed:
feat(ChessGameInfoRepository) : GameInfo 를 담당하는 레포지터리 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
refactor(Rank) : Rank 숫자가 아닌 문자열로 변경

Point 도 같이 변경
- 이에 따른 테스트 코드 변경

youngsu5582 committed:
feat(Color,File,PieceStatus) : DB 에 저장할 value 추가 , getter 및 from 추가

youngsu5582 committed:
feat(PieceIdGenerator) : Piece Id 를 생성해주는 기능 구현

youngsu5582 committed:
feat(Piece) : 기물들에 PieceId 추가

youngsu5582 committed:
feat(PieceEntityRepository) : PieceEntity 를 담당하는 레포지터리 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(ChessBoard) : 멤버 변수에 gameId 추가

- 이에 따른 코드들 변경

youngsu5582 committed:
feat(ChessService) : ChessBoard - Repository 로직 담당하는 기능 구현

youngsu5582 committed:
feat(ChessController) : 방번호 입력하는 기능 구현

youngsu5582 committed:
feat(ChessGameInfoRepository) : GameInfo 삭제하는 기능 로직 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(PieceEntityRepository) : gameId 에 따른 모든 기물들 삭제하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 킹 기물인지 확인하는 기능 구현

youngsu5582 committed:
refactor(ChessBoard) : 턴 교체 기능 외부로 분리

youngsu5582 committed:
refactor(ChessGameInfo) : 색깔 반전후 객체 생성해 반환하게 변경 , 아이디 - 색깔 에 대한 게임정보 캐싱

이에 따른 기본 생성자 -> valueOf 로 변경

youngsu5582 committed:
refactor(ChessService) : 이동 기능 분리 , 기능 수행후 ChessBoard 반환하게 변경

youngsu5582 committed:
fix(ChessGameInfo,PieceEntityMemoryRepository) : Map 반복 중 수정하는 코드 변경

youngsu5582 committed:
feat(CheckMateException) : 왕이 잡으면 Controller 에서 Catch 하게 변경

youngsu5582 committed:
feat(ChessController) : 컨트롤러가 CheckMateException catch 하게 변경 , gameId 를 통해 로직 수행하게 변경

커맨드 입력 예시 알려주게 변경

youngsu5582 committed:
feat(ConnectionGenerator) : DB 연결하여 커넥션을 반환하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(ChessGameInfoSqlRepository) : ChessGameInfoRepository 를 DB로 기능 구현

- 기존 테스트 코드를 활용해 검증

youngsu5582 committed:
feat(PieceEntitySqlRepository) :
PieceEntityRepository 를 DB로 기능 구현

- 기존 테스트 코드를 활용해 검증

youngsu5582 committed:
refactor(all) : 패키지 재 분리

youngsu5582 committed:
feat(Docker) : DB 구동해주는 도커 내용 추가

docker-compose 를 통해 구동
- 테이블 초기화하는 create sql 추가
- 요구사항에 추가
.gitignore 에 docker/db/mysql 에 추가

youngsu5582 committed:
refactor(Direction) : canMoveDirection -> canMoveOneDirection 으로 명확하게 명시

이에 따른 코드들 수정

youngsu5582 committed:
refactor(GameEndException) : CheckMateException -> GameEndException 으로 변경

이에 따른 코드들 수정

youngsu5582 commented on src/main/java/domain/chess/piece/PieceChecker.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class PieceChecker implements PieceCheckable {
```
해당 부분은 Piece 가 Pieces 를 받아서
양방향 의존관계를 피하기 위해 만들어진 도메인 입니다.
List<Piece> 를 가지는 건 동일하나 각자 맡은 역활을 다르게 했는데요!
이게 제이가 의도한 설명이 맞을까요??

JunHoPark93 replied:
piece와 pieces의 단방향 구현을 하기위해서 추가적인 클래스가 나온거같은데 저는 추가적인 클래스를 의도한건 아니긴하지만, 기존 구조에서 추가적인 클래스 없이 리팩토링하기 힘든 구조였을까요?

youngsu5582 replied:
### [[추가 질문]]
추가적인 클래스 없이를 생각해보면,
검증 로직들 중 공통적인 기물들 끼리 분류해서
그 로직들 끼리는 List<Piece> 를 받아서 수행하는 식으로 할 꺼 같은데
제이의 의도가 이거일까요?? 🙇‍♂️

JunHoPark93 replied:
pieces내부로직을 말씀하시는거죠? 넵 구현방향이 어떻든 추가적인 클래스가 나오지 않는 방향을 의도했습니다

youngsu5582 commented on src/main/java/domain/chess/ChessBoard.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.List;
+import java.util.Optional;
+
+public class ChessBoard {
```
ChessBoard 는
DB가 추가됨에 따라 , 이제 게임 종료 , 상태 관리 등 모든 로직을 담당하는게 아니라
생성이 타당한지 & 이동이 타당한지 & 이동후 기물의 위치만 반환합니다.

상태 관리는 -> Repository & DB
게임 종료 같은 행위 관리는 -> Service
가 담당하는게 맞다고 생각해서 분리 했는데요 ,,!

DB 를 접목시키는게 처음이다 보니
해당 기준이 맞는지 확신이 없는거 같습니다 😭
어떨까요 ...?

JunHoPark93 replied:
넵 역할은 잘 나눠주신거같아요!
정답은 없지만 chessboard하나에 많은 책임이 들어있는것보단 좋은거같습니다.

JunHoPark93 replied:
"상태 관리는 -> Repository & DB" 라고하신부분에 체스관련 로직을 의도하신건아니죠?

youngsu5582 replied:
데이터 저장 & 관리의 관점에서 Repository & DB 를 생각했습니다!
관련 로직은 Domain 이 담당하나 , 도메인 과 DB및 엔티티 는 가능한 또 분리하는게 좋다고 생각했습니다!

youngsu5582 reviewed (commented):
코드적인 부분에서 궁금증은 File Changes에 남겼습니다!
다른 부분도 피드백 주거나 설계에서 문제가 되는 부분이 있다면 과감하게 리뷰 주셔도 감사합니다! 🙂 🙇‍♂️

youngsu5582 committed:
refactor(ChessGameInfoEntity) : 상수 네이밍 변경 & 최상단 위치

youngsu5582 committed:
refactor(ChessService,PieceChecker) : 함수형 으로 리팩토링

youngsu5582 committed:
refactor(ChessController) : gameId 초기화 하는 로직 함수로 분리

youngsu5582 committed:
refactor(Application) : connectionGenerator 같은 객체로 사용

JunHoPark93 commented on src/test/java/fake/ChessGameInfoMemoryRepository.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class ChessGameInfoMemoryRepository implements ChessGameInfoRepository {
```
> 1. Fake Repository 구현
> 4단계 미션을 진행하며
> 우선 interface 와 Map 을 활용하여 Fake Repository 를 먼저 구현한 후
> 실제 DB와 연동하는 Repository 를 만드는 식으로 구현을 하였습니다.
> 
> 이유로는
> 
> 내가 필요한 데이터 , 메소드 가 뭔지 설계 & 동작하는 원리를 알기 위해서
> 먼저 작성하고 통과한 테스트 코드를 기반으로 DB를 하나씩 확인 가능하기 위해서
> 가 있었습니다
> 구현하며 느낀점은
> 똑같은 작업이 반복은 되나 , 편안함을 느꼈습니다.
> 해당 내용을 기반으로 비즈니스 로직을 완료함 후,
> 디비로 교체하여 사용하니 문제없이 바로 교체가 가능했기 때문인데요,
> 실무에서도 이렇게 FakeRepository 를 통한 접근 방식이 자주 사용되는지 궁금합니다!

이렇게 쓰기도하지만 저는 주로 mock객체를 활용했었고 통합테스트때에는 실제 repository 객체를 사용해서 테스트하도록 했었네요. 어떤부분을 테스트 하냐에 따라 달라지기 때문에 필요에 맞게 해주시면 될거같습니다.

youngsu5582 replied:
상황에 맞춰서 적용을 하면 된다 이군요. 감사합니다! 😊

JunHoPark93 commented on src/main/java/domain/chess/piece/Color.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Arrays;
+
+public enum Color {
+    BLACK("black"),
+    WHITE("white");
```
> 해당 미션을 구현하며 디비 설계를 했습니다.
> 그리고 , DB 에는 객체가 아닌 문자열을 저장할 수 밖에 없습니다.
> DTO 를 변화시켜주듯이 Mapper 를 사용할까 생각했으나
> 
> 다른 클래스에서 관리하는게 아닌 해당 값은 해당 값과 매칭 되는 Enum 이 가지고 있는게 더 적합하다 생각하여 문자열 과 문자열을 통해 Enum을 반환하는 로직을 만들었습니다.
> DB 의 값과 Enum 내 문자열을 매칭 시키는게 일반적인지 궁금합니다!

넵 enum 문자열 그대로 들어가는게 일반적입니다. (오히려 상수형태로 들어가는게 문제가 되는 소지가 있습니다)

youngsu5582 replied:
### [[추가 질문]]
상수 형태로 들어가는게 문제가 되는 이유로는
의도치 않은 값이나 상수 순서에 의존이 되게 때문 일까요??! 🤩

JunHoPark93 replied:
의도치 않은 값이라기보다는 말씀주신 순서문제와 조회시 해당 값이 어떤 의미를 갖는지 한번에 알 수가 없다는 불편함이 있습니다.

JunHoPark93 commented on src/main/java/domain/chess/exception/CheckMateException.java:
```diff
@@ -0,0 +1,9 @@
+
+    public CheckMateException(final String input) {
+        super(String.format(MESSAGE, input));
+    }
```
> 3단계 점수 및 게임 종료 로직을 구현하며
> 왕이 잡히면 다음 입력 로직 이나 , 비즈니스 로직을 수행하지 않고
> 바로 종료가 되게 하고 싶었습니다.
> 
> 그렇기에 GameEndException 같이 Exception 을 발생시키고 , catch 하는 방법을 사용했습니다.
> 이때 든 생각으로
> 객체지향의 세계에서는 결국 객체간 메시지들을 보내 단방향으로 오가야 하나
> Exception 을 비즈니스 로직에 사용하여 의도적으로 로직을 끊는 일종의 C언어의 goto의 느낌을 받았습니다.
> 
> 또 한편으론 , 체스 게임에선
> 왕이 잡히면 게임이 끝나는 것이니 GameEndException 이나 CheckMateException 을 발생 시켜도 괜찮지 않을까?
> 라는 생각도 들기도 합니다.
> 
> 비즈니스 로직 & 로직 중 의도적으로 Exception 을 발생시켜 로직을 수행해도 괜찮은지 궁금합니다!
> ( 아니면 , Exception 을 통해 비즈니스 로직 & 게임을 종료 시켜도 되는지...! )

왕이 잡히는것이 예외라고 생각하신걸까요?
게임이 끝나려면 왕이 잡혀야합니다. 위 로직대로라면 게임이 끝나려면 예외가 발생해야한다는 말과 같습니다. 게임의 흐름을 다시 생각해보면 좋을거같아요.

JunHoPark93 replied:
exception처리는 의도적으로 로직을 끊는게 아니라 끊어지지않게 처리를 하는것에 더 가깝습니다. 자바에서의 exception에 대해서 좀더 찾아보시면 좋을거같아요!

youngsu5582 replied:
오히려, 다른 관점으로 생각을 하고 있었네요..!
Exception을 통해 다른 곳에서 catch 하여, 로직을 제어하거나 하는게 아닌
의도한 로직 대로 흐르게 바로 catch 하고, 다음 로직 부터는 의도한 로직 대로 동작하게 해주는 역활 이군요!
저도 게임이 끝나려면 예외가 발생해야 한다는 부분에서 계속 의구심이 들었습니다!

JunHoPark93 replied:
넵 자료찾아보시면 많이 나와있으니 꼭 학습하고 넘어가시길 바랍니다!

JunHoPark93 commented on src/main/java/domain/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,95 @@
+
+public class ChessController {
+    private final ChessService chessService;
+    private Integer gameId;
```
꼭 필요한 인스턴스 변수일까요?

youngsu5582 replied:
처음 의도한 접근방식은

사용자가 매번 입력하는 걸 방지하기 위해서 Browser 에서 Cache의 느낌으로 사용하자!
였습니다.
하지만 그건 나중에 웹단이나 애플리케이션에서 신경 써야할 부분이지, Controller 가 신경 써야할 부분이 아님을 깨달았습니다!

gameId 를 삭제 + Exception을 발생 시키지 않는 방식으로 할 꺼같습니다!

JunHoPark93 commented on src/main/java/domain/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,95 @@
+                }
+                InputView.clear();
+                OutputView.printException(exception.getMessage());
+            }
```
위에 코멘트 드린것처럼 예외처리부분을 개선해보시면 좋을거같아요~

JunHoPark93 commented on src/main/java/domain/chess/repository/ChessGameInfoSqlRepository.java:
```diff
@@ -0,0 +1,84 @@
+        if (resultSet.next()) {
+            final var id = resultSet.getInt("game_id");
+            final var color = resultSet.getString("color");
+            return Optional.ofNullable(new ChessGameInfoEntity(id, Color.from(color)));
```
항상 null이 아니므로 of로 써도될거같네요

youngsu5582 replied:
resultSet 이 next 를 true 로 반환 했을 경우
id 와 color를 못받아올 경우는 없을테니까 of 로 선언해도 괜찮을 거 같습니다!

JunHoPark93 commented on src/main/java/domain/chess/service/ChessService.java:
```diff
@@ -0,0 +1,91 @@
+        final List<PieceEntity> pieceEntities = this.pieceEntityRepository.findAllByGameId(chessGameInfoEntity.chessGameId());
+        return new ChessBoard(new Pieces(pieceEntities.stream()
+                                                      .map(PieceEntity::toPiece)
+                                                      .toList()), chessGameInfoEntity.color(), chessGameInfoEntity.chessGameId());
```
코드를 좀 더 간결하게 개선해볼 수 있을까요?

youngsu5582 replied:
단순히 생각나는 방법으로는
Pieces 를 지역변수로 선언해서 사용하는 방법 밖에 떠오르지 않는데
### [[추가질문]]
ChessBoard 에 정적 팩토리 메소드를 만들어서 
fromEntity() 이런식으로 만드는 방식이 괜찮은지 궁금합니다!

꺼리는 이유로 생각한다면
결국 도메인이 Entity 클래스를 알게된다..? ( 도메인에서 Entity로 변환되는건 자연스러운 단방향 흐름 )
한편으론,
변환하는 로직 이나, 특수한 로직이 있는게 아닌 단순 받아서 생성하는것이니 상관없다?

의 생각이 떠오릅니다!

JunHoPark93 replied:
단순히 new Pieces하는 부분을 private method로 만들기만해도 한 메서드의 크기가 줄어들거에요. 말씀주신 mapper형태의 로직은 선택입니다.

JunHoPark93 commented on src/main/java/domain/chess/service/ChessService.java:
```diff
@@ -0,0 +1,91 @@
+        if (chessGameInfo.isPresent()) {
+            return getExistChessBoard(chessGameInfo.get());
+        }
+        return createNewChessBoard(gameId);
```
Optional의 map을 활용해서 간결하게 만들어볼 수 있을거에요

youngsu5582 replied:
Optional 을 제대로 사용해본건 , 이번 미션이 처음이라 잘 몰랐는데
Optional 도 간결하게 표현이 가능하네요!! ( 안그래도 쓰면서, Optional 은 꼭 이렇게 써야하나?? 에 대한 불신이 있었습니다.. )
```java
return this.chessGameInfoRepository.getChessGameInfoWithGameId(gameId)
                                   .map(this::getExistChessBoard)
                                   .orElseGet(() -> createNewChessBoard(gameId));
```
해당 코드를 의도한게 맞을까요??!

JunHoPark93 replied:
넵 맞습니다 optional 처리방법에 대한 다양한 내용들은 자료 찾아보시면 많이 나와있으니 학습하시고 넘어가시길 바랍니다!

JunHoPark93 commented on src/main/java/domain/chess/piece/Piece.java:
```diff
@@ -1,23 +1,39 @@
     protected Piece(final Point point, final Color color) {
         this.point = point;
         this.color = color;
+        this.pieceId = PieceIdGenerator.generate();
```
uuid로 생성하신 이유가 있을까요?

youngsu5582 replied:
### [[느낀점]]

해당 부분은 , 기존에는 gameId , point , color , kind 들을 통해 식별자를 만들려고 했으나 ,
로직상 기물의 point 가 변경되는 점 때문에 식별할 수 없게 되어 고유 식별자 가 필요하게 되었습니다.

이때 숫자와 uuid 중 뭘 선택 하지에 대해 생각하던 중
디비에서 auto increment 를 해서 디비 에서 생성하는게 아닌
서버에서 id를 할당해주므로 uuid 를 통해 여러 서버에서 도 id가 중복되지 않게하자!
라고 생각했는데 , 제이의 의견에 대해 답변을 작성하며 생각해보니
숫자여도 크게 상관이 없을꺼 같긴하네요..!
( PieceEntity 를 통해 insert 시 자동 id 생성 -> select 시에는 자동 생성된 id 도 같이 받아오는 형식 )

JunHoPark93 replied:
넵 해도 상관없지만 현재 상황에서는 굳이 필요없을거같아보어셔요!

JunHoPark93 commented on src/main/java/domain/chess/piece/PieceChecker.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    public boolean isEmptyPoint(final Point point) {
+        final Optional<Piece> optionalPiece = this.findPieceWithPoint(point);
+        return optionalPiece.isEmpty();
+    }
```
불필요한 로컬변수를 삭제해보면 좋을거같아요. 다른부분들도 확인해주세요!

`return findPieceWithPoint(point).isEmpty();`

JunHoPark93 commented on docker/db/init/chessGameInfo_create.sql:
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE ChessGameInfo (
+    game_id INT PRIMARY KEY ,
+    color ENUM('black','white') NOT NULL
```
다른 형태들이 추가가될때 수정해야되는 부분들이 많아져서 varchar로 하는게 좋을거같긴한데 참고만 해주세요!

youngsu5582 replied:
DB 단 에서는 오히려 엄격히 규제하는 enum 이 안좋을수도 있겠네요!

JunHoPark93 commented on src/main/java/domain/chess/piece/PieceChecker.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    public boolean isNotEqualColorInPoint(final Color color, final Point point) {
+        final Optional<Piece> optionalPiece = this.findPieceWithPoint(point);
+        if (optionalPiece.isEmpty()) {
+            return false;
+        }
+        final Piece toPiece = optionalPiece.get();
+        return !toPiece.isEqualColor(color);
```
전체적으로 optional처리하는 부분들도 개선할 수 있을지 확인해주시면 좋을거같아요.

youngsu5582 replied:
불필요한 분리였네요!

JunHoPark93 commented on src/main/java/domain/chess/piece/Pieces.java:
```diff
@@ -21,14 +21,25 @@ public void move(final Piece piece, final Point point) {
 
-    public boolean containPieceWithPoint(final Point point) {
-        return findPieceWithPoint(point).isPresent();
+    public double getScore(final Color color) {
+        return value.stream()
+                    .filter(piece -> piece.isEqualColor(color))
+                    .mapToDouble(piece -> piece.getScore(value))
+                    .sum();
     }
```
> 현재의 요구사항은
> 일반적으로 다른 기물 들은 단순 상수 점수값을 반환하나
> 폰은 같은 랭크 에 다른 아군 폰이 있을 시 0.5점을 반환하여 추가된 로직이 포함되어 있습니다
> 처음에는 폰은 특별하게 분리를 하여 로직을 구분을 할까 했지만
> 
> 다른 기물들 에서 이와 유사한 로직을 해야할 시 또 isXX 를 만들어 로직을 분리해야 한다는 점
> Pawn 이 List 구문이 필요 없어진다면 삭제하기 용이하다는 점
> 이러한 점들 때문에 사용하지 않지만 기물들 getScore 에 List 를 넣었습니다!
> 위 접근 방식이 타당 할지 궁금합니다 제이!

죄송하지만 제가 질문을 잘 이해를 못한거같습니다. 이 부분이 맞나요?
pawn이 list구문이 필요없어지면 삭제하기 용이하다는게 어떤 말씀이실까요?

youngsu5582 replied:
### [[추가 질문]]

제가  너무 추상적으로만 설명했네요,, 죄송합니다 😭

현재의 요구사항에서
`pawn의 기본 점수는 1점이다. 하지만 같은 세로줄에 같은 색의 폰이 있는 경우 1점이 아닌 0.5점을 준다.`
이라는 pawn 만 특수한 점수 요구사항을 가지고 있습니다.

이에 따라 Pawn을 제외한 나머지 Piece 들은 `getScore()`를 가지고 
Pawn만 `getScore(List<Piece> pieces)`을 가지게 할지에 대해서 고민했습니다.

하지만 이러면 결국 다형성 을 해친다고 생각하여
다른 Piece 들은 입력받은 매개변수 를 사용하지 않더라도
`public abstract double getScore(List<Piece> pieces);`
이렇게 상위 Class 인 Piece 에서 통일되게 만들었습니다.

1. 나중에 다른 기물에 대한 요구사항이 생긴다면? -> 해당 기물들을 통해 로직 수행
2. 나중에 Pawn에 대한 요구사항이 사라진다면? -> 매개변수 삭제 & 폰 로직 수정

이렇게 확장이 용이하다고 생각하는데
해당 접근 방식이 타당한지에 대해 물어봤습니다!
( 제가 설명을 잘 못해서,, 이해 안되거나 제 의견이 궁금하시면 다시 Comment 나 DM 주셔도 괜찮습니다! ☺️ )

JunHoPark93 replied:
아 넵 폰때문에 구조가 깨지는것을 고려하신판단이군요. 폰만을 위한 로직을 넣거나 다 통일하거나 해야하는데 어느하나 다 만족시킬수 없으니 작성주신 list방식이 나은거같아요. 폰에 대한 요구사항이 사라질일은 거의 없을거기도하구요!

JunHoPark93 commented on src/main/java/domain/chess/service/ChessService.java:
```diff
@@ -0,0 +1,91 @@
+
+
+    public ChessBoard moveChessBoard(final int gameId, final RouteDto routeDto) {
+        final var chessBoard = getExistChessBoard(gameId);
+        final var optionalPiece = chessBoard.findPiece(routeDto.getEndPoint());
+
+        final var piece = chessBoard.move(routeDto.getStartPoint(), routeDto.getEndPoint());
+        optionalPiece.ifPresent(endPointPiece -> processDelete(chessBoard, endPointPiece));
+
+        processMove(chessBoard, piece);
+        return chessBoard;
+    }
```
선언된 로컬변수를 바로 다음 라인에 사용하도록 줄 변경을 해보면 불필요한 선언을 줄일 수 있을거에요!

youngsu5582 replied:
굳이 optional 과 ifPresent 구문을 분리할 필요 없이 바로 사용이 가능하네요...! 감사합니다!! ☺️

JunHoPark93 commented on src/test/java/fake/PieceEntityMemoryRepository.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public List<PieceEntity> findAllByGameId(final int gameId) {
+        return this.map.entrySet()
```
map의 value만을 다루기때문에 map.values()로 시작해도 될거같아요

youngsu5582 replied:
불필요한 entrySet 이였네요,, 감사합니다!

JunHoPark93 commented on src/main/java/domain/chess/piece/kind/Rook.java:
```diff
@@ -0,0 +1,45 @@
+        return PieceStatus.ROOK;
+    }
+
+    public boolean canMove(final Point movePoint, final List<Piece> pieceList) {
```
기물각자가 움직임을 판단할 수 있어서 좋네요 👍🏼

youngsu5582 replied:
저도 기물 스스로가 움직임을 판단하는게
다형성을 해치지 않아서 좋은거 같습니다!!

JunHoPark93 reviewed (changes requested):
안녕하세요 조이썬~
마지막 미션 하느라 고생많으셨습니다.
코멘트 및 질문주신 부분들 남겨두었는데 확인 후 의견 및 반영부탁드릴게요!

youngsu5582 committed:
refactor(Pieces) : 왕을 찾는 로직 왕의 숫자 찾는 로직으로 변경

불필요한 Optional 사용 함수 제거 목적 위함
ChessBoard 에서도 검증 방식 변경

youngsu5582 committed:
refactor(Repository) : Optional.nullable 이 아닌 of 로 변경

절대 Null 이 될 수 없는 상황이므로 of 사용
[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/783#discussion_r1546379385)

youngsu5582 committed:
refactor(ChessService,PieceChecker) : 불필요한 지역 변수 제거

불필요 하게 지역 변수 받고, return 하는 코드 변경
[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/783#discussion_r1546365789)

youngsu5582 committed:
feat(ChessController,ChessService,GameEndException) : gameId 제거, 예외 발생 제거

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/783#discussion_r1546353165)

youngsu5582 committed:
chore(sql) : Enum 부분 varchar 로 변경

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/783#discussion_r1546345806)

youngsu5582 commented:
제이!
좀 늦게 수정해서 올리는거 같아서 죄송하네요,,

몇 가지 부분에서 고민한 부분이 있어서 그렇습니다!
바로 시작하겠습니다! 🫡

---
#### Optional.get 의 사용
```java
public Piece move(final Point startPoint, final Point endPoint) {
    validatePoint(startPoint, endPoint);
    final Optional<Piece> optionalPiece = this.pieces.findPieceWithPoint(startPoint);
    final Piece piece = optionalPiece.get();
    pieces.move(piece, endPoint);
    return piece;
}
```
해당 코드에서, piece 는 단순히 optionalPiece.get 을 하는데 이 부분은
```java
    private void validatePoint(final Point startPoint) {
        final Optional<Piece> optionalPiece = this.pieces.findPieceWithPoint(startPoint);
        if (optionalPiece.isEmpty()) {
            throw new IllegalArgumentException(String.format("%s에는 기물이 없습니다", startPoint));
        }
        optionalPiece.ifPresent(this::validateDifferentColor);
    }
```
이렇게 validatePoint 부분에서 기물이 있는지 검증을 합니다.
이때 get 을 피하기 위한 방법으로

1. validate 부분에서 검증과 동시에 piece 를 반환한다.
2. 위에서 검증하고, 다른 함수인 getPieceWithPoint()를 만든다
```java
getPieceWithPoint(Point point){
    return value.stream()
            .filter(piece -> piece.isEqualPoint(point))
            .findAny();
            .orElseThrow(...);
``` 
3. validatepoint 부분은 분리한다

이 두 방식을 생각했으나, 근본적인 해결책이 아니라 생각해서 바로 .get 을 사용했는데 sonarlint 검사에서 경고가 나와서
이런 기존의 검사가 명확한 경우에도 해야하는가? 에 대한 궁금증이 생겼습니다.
( 이것도 기존에 제가 질문했던, 도메인 단 교차검증과 같은 느낌 일까요?? )

#### 도메인 과 엔티티의 분리?

4단계 미션을 진행하며 느낀점은
도메인 과 DB에 저장하는 엔티티를 어떻게 분리해야 하고, 또 어떻게 합쳐야 하지...? 였습니다

```java
public abstract class Piece {
    protected Point point;
    protected final Color color;
    private final String pieceId;
    ...
}
```
해당 도메인에 대해서
```
public record PieceEntity(String point, String color, String kind, String pieceId, int gameId){
...
}
```
이렇게 도메인을 전부 변환한 Entity를 만들었습니다.
엔티티도 결국 View 처럼 도메인 관점에서 보면 외부의 관점이 될 수 있을거 같아서 
도메인의 변화에 의존이 되지 않게 하기 위해 순수 DB에 저장하는 값으로 다 변환을 하자! 라고 생각을 했는데요.
```java
public static PieceEntity fromPiece(final Piece piece, final int gameId) {
    final String point = piece.toPointEntity();
    final String color = piece.toColorEntity();
    final String kind = piece.toKindEntity();
    final String pieceId = piece.getPieceId();

    return new PieceEntity(point, color, kind, pieceId, gameId);
}
```
이렇게 Entity 가 Mapper 의 역활 이 된 거 같아서 해당 방식이 맞나?에 대해서 의문이 들었습니다
Entity 를 처음 써보는데 `도메인 변경에 영향을 받지 않고, 순수 로직이 없고 DB에 저장하기 위한 객체`
의 접근 방식이 맞는지 궁금합니다!

---

추가적인 부분은 File Changed 부분에 남겼습니다!
( 제이의 의견이 더 듣고 싶은 부분은 `[[느낀점]]` 으로,
  제이에게 다시 질문한 부분은 `[[추가 질문]]`으로 마크 해놨습니다! - 검색해서 하시면 될 거 같아요! 🙂)

마지막 까지 잘 부탁드립니다!! 🫡🫡

JunHoPark93 commented:
## 1번질문
> 이런 기존의 검사가 명확한 경우에도 해야하는가? 에 대한 궁금증이 생겼습니다.

제가 질문을 제대로 파악을 못한거같긴한데요, optional처리개선이 목적이신거같아서 대략 아래와같이 시도해보시면 될거같아요.
```java
this.pieces.findPieceWithPoint(startPoint).ifPresentOrElse(
                this::validateDifferentColor,
                () -> {throw new IllegalArgumentException(String.format("%s에는 기물이 없습니다", startPoint));}
        );
```

## 2번질문

> 엔티티도 결국 View 처럼 도메인 관점에서 보면 외부의 관점이 될 수 있을거 같아서
> 도메인의 변화에 의존이 되지 않게 하기 위해 순수 DB에 저장하는 값으로 다 변환을 하자! 라고 생각을 했는데요.

외부의 관점이라는게 어떤 말씀이실까요? 

정답은 없고 설계하기 나름입니다. 다만 entity클래스와 domain클래스를 나누고자하는 목적이 있다면 말씀주신 방향대로 하면 됩니다. 편의상 나누지않고 쓰는경우도 존재하지만, 이렇게 나눴을때 장/단점이 있을텐데요. 해당부분은 스스로 찾아보시면 좋을거같아요. 
위 부분은 꼭 지금 답을 내지 않아도 좋습니다. 프로젝트하실때에도 같은 고민을 또 하시게될거에요!

JunHoPark93 reviewed (approved):
안녕하세요 조이썬!
리뷰 반영하느라 고생많으셨어요.
질문 주신 부분은 제가 코멘트 남겨두었는데 해결이 되지않았다면 dm부탁드릴게요!
다음 미션도 화이팅입니다!
